src/execution/execute.ts

```diff
@@ -286,87 +272,321 @@ export function buildExecutionContext(
+    typeResolver: options.typeResolver ?? defaultTypeResolver,
+    subscribeFieldResolver:
+      options.subscribeFieldResolver ?? defaultFieldResolver,
   };
+  return { result };
 }
 
 /**
- * Implements the "Executing operations" section of the spec.
+ * Constructs a ExecutableRequest object
  */
-function executeOperation(
-  exeContext: ExecutionContext,
-  operation: OperationDefinitionNode,
-): PromiseOrValue<ObjMap<unknown>> {
-  const rootType = exeContext.schema.getRootType(operation.operation);
-  if (rootType == null) {
-    throw new GraphQLError(
-      `Schema is not configured to execute ${operation.operation} operation.`,
-      { nodes: operation },
-    );
-  }
-
-  const rootFields = collectFields(
-    exeContext.schema,
-    exeContext.fragments,
-    exeContext.variableValues,
-    rootType,
-    operation.selectionSet,
+export function makeExecutableRequest(
+  schema: GraphQLSchema,
+  document: DocumentNode,
+  operationName: Maybe<string>,
+  options: GraphQLExecutionPlanOptions = {},
+): ResultOrGraphQLErrors<ExecutableRequest> {
+  const makeExecutionPlanReturn = makeExecutionPlan(
+    schema,
+    document,
+    operationName,
+    options,
   );
-  const path = undefined;
 
-  const { rootValue } = exeContext;
+  if (makeExecutionPlanReturn.errors !== undefined) {
+    return makeExecutionPlanReturn;
+  }
+  const executionPlan = makeExecutionPlanReturn.result;
 
-  switch (operation.operation) {
+  switch (executionPlan.operation.operation) {
     case OperationTypeNode.QUERY:
-      return executeFields(exeContext, rootType, rootValue, path, rootFields);
+      return { result: new ExecutableQueryRequestImpl(executionPlan) };
     case OperationTypeNode.MUTATION:
-      return executeFieldsSerially(
-        exeContext,
-        rootType,
-        rootValue,
-        path,
-        rootFields,
-      );
+      return { result: new ExecutableMutationRequestImpl(executionPlan) };
     case OperationTypeNode.SUBSCRIPTION:
-      // TODO: deprecate `subscribe` and move all logic here
-      // Temporary solution until we finish merging execute and subscribe together
-      return executeFields(exeContext, rootType, rootValue, path, rootFields);
+      return { result: new ExecutableSubscriptionRequestImpl(executionPlan) };
+  }
+}
+
+class CoercedVariableValues {
+  coercedValues: { [variable: string]: unknown };
+
+  constructor(objMap: { [variable: string]: unknown }) {
+    this.coercedValues = objMap;
+  }
+}
+
+export type ExecutableRequest =
+  | ExecutableQueryRequest
+  | ExecutableMutationRequest
+  | ExecutableSubscriptionRequest;
+
+export interface ExecutableQueryRequest {
+  operationType: OperationTypeNode.QUERY;
+
+  coerceVariableValues: (
+    rawVariableValues: Maybe<{ readonly [variable: string]: unknown }>,
+  ) => ResultOrGraphQLErrors<CoercedVariableValues>;
+
+  /**
+   * Implements the "ExecuteQuery" algorithm described in the GraphQL specification.
+   */
+  executeOperation: (
```

Comment:
see above, I like how executeOperation now actually executes every operation (whereas previously the term was used for retrieving the data-only for queries/mutations). I just think we can deprecate the execute stage of the pipeline and utilize executeOperation, which would take a set of cacheable parameters (schema/document) as well as the 3 arguments here).

Reply:
> I just think we can deprecate the execute stage of the pipeline and utilize executeOperation, which would take a set of cacheable parameters (schema/document) as well as the 3 arguments here).

We can do that as an intermediate step.
Just to clarify you are proposing using `WeakMap` internally to cache stuff based on schema/document/operationName as you do in `graphql-executor`?
Or are you proposing to use externally catchable `ExecutionPlan`as a replacement for schema/document/operationName?

Reply:
The latter.

Reply:
But as I wrote above, although the beginnings of #3314 r here, there is very little that is being actually done and stored within the plan (afaict) besides storing the a validated root type. In terms of the exported interface, it also would not be enough (afaict) to build a customized stitching executor which needs access to later steps in the execution algorithm. Although I think I see the direction, in terms of functionality, I am not sure this as yet brings enough minimum value to be worth the complexity.

Reply:
I was constrained within graphql-executor because I wanted it to be a drop in replacement for graphql-js. So there, I used a simple memoization strategy that assumed if you have cached a document, everything about it should be cached.

If we are talking about changing the graphql-js api - I think exporting a few more low level building blocks is the low-hanging fruit.